src/golang/lib/workflow/operator/base.go

```diff
@@ -123,49 +126,30 @@ func (bo *baseOperator) fetchExecState(ctx context.Context) *shared.ExecutionSta
 			"Unable to read operator metadata from storage. Operator may have failed before writing metadata.",
 		)
 	}
+
 	return &execState
 }
 
-// GetExecState takes a more wholelistic view of the operator's status than the job manager does,
-// and can be called at any time. Because of this, the logic for figuring out the correct state is
-// a little more involved.
-func (bo *baseOperator) GetExecState(ctx context.Context) (*shared.ExecutionState, error) {
-	if bo.jobName == "" {
-		return nil, errors.Newf("Internal error: a job name was not set for this operator.")
-	}
-
-	status, err := bo.jobManager.Poll(ctx, bo.jobName)
-	if err != nil {
-		// If the job does not exist, this could mean that
-		// 1) it is hasn't been run yet (pending),
-		// 2) it has run already at sometime in the past, but has been garbage collected
-		// 3) it has run already at sometime in the past, but did not go through the job manager.
-		//    (this can happen when the output artifacts have already been cached).
-		if err == job.ErrJobNotExist {
-			// Check whether the operator has actually completed.
-			if utils.ObjectExistsInStorage(ctx, bo.storageConfig, bo.metadataPath) {
-				return bo.fetchExecState(ctx), nil
-			}
-
-			// Otherwise, this job has not run yet and is in a pending state.
-			return &shared.ExecutionState{
-				Status: shared.PendingExecutionStatus,
-			}, nil
-		} else {
-			// This is just an internal polling error state.
-			return unknownSystemFailureExecState(err, "Unable to poll job manager."), nil
-		}
+// updateExecState and merge timestamps with current state based on the status of the new state.
+// Other fields of bo.execState will be replaced.
+func (bo *baseOperator) updateExecState(execState *shared.ExecutionState) {
+	now := time.Now()
+	// copy current timestamps to merge these time
+	execTimestamps := bo.execState.Timestamps
+	if execState.Status == shared.FailedExecutionStatus {
+		execTimestamps.FailedAt = &now
+	} else if execState.Status == shared.SucceededExecutionStatus {
+		execTimestamps.SucceededAt = &now
+	} else if execState.Status == shared.RunningExecutionStatus {
+		execTimestamps.RunningAt = &now
 	} else {
-		// The job just completed, so we know we can fetch the results (succeeded/failed).
-		if status == shared.FailedExecutionStatus || status == shared.SucceededExecutionStatus {
-			return bo.fetchExecState(ctx), nil
-		}
-
-		// The job must exist at this point, but it hasn't completed (running).
-		return &shared.ExecutionState{
-			Status: shared.RunningExecutionStatus,
-		}, nil
+		// This method shouldn't get called under other states.
+		// As an additional safeguard, we do nothing for other states if it's called.
+		return
```

Comment:
Can we return an error here? We should enforce internal assumptions as much as possible.